sihl/src/database_migration.ml

```diff
@@ -220,10 +221,68 @@ struct
     execute steps
   ;;
 
+  let migrations_status () =
+    let open Lwt.Syntax in
+    let* migrations_states = Repo.get_all (table ()) in
+    let migration_states_namespaces =
+      migrations_states
+      |> List.map (fun migration_state ->
+             migration_state.Database_migration_repo.Migration.namespace)
+    in
+    let registered_migrations_namespaces =
+      Map.to_seq !registered_migrations |> List.of_seq |> List.map fst
+    in
+    let namespaces_to_check =
+      List.concat
+        [ migration_states_namespaces; registered_migrations_namespaces ]
+    in
+    Lwt.return
+    @@ List.map
+         (fun namespace ->
+           let migrations = Map.find_opt namespace !registered_migrations in
+           let migration_state =
+             List.find_opt
+               (fun migration_state ->
+                 String.equal
+                   migration_state.Database_migration_repo.Migration.namespace
+                   namespace)
+               migrations_states
+           in
+           match migrations, migration_state with
+           | None, None -> namespace, None
+           | None, Some migration_state ->
+             ( namespace
+             , Some (-migration_state.Database_migration_repo.Migration.version)
+             )
+           | Some migrations, Some migration_state ->
+             let unapplied_migrations_count =
+               List.length migrations
+               - migration_state.Database_migration_repo.Migration.version
+             in
+             namespace, Some unapplied_migrations_count
+           | Some migrations, None -> namespace, Some (List.length migrations))
+         namespaces_to_check
+  ;;
+
+  let pending_migrations () =
+    let open Lwt.Syntax in
+    let* unapplied = migrations_status () in
+    let rec find_pending result = function
```

Comment:
Could this be replaced with `List.filter_map snd unapplied |> List.filter (fun mig -> snd mig > 0)`

`filter_map` should get rid of all the `None` and should automatically unwrap all `Some`

Reply:
Unapplied has type `(string * int option) list` so I don't see how this works since I need the namespaces as well in the end.
I prefer local `let rec ...` over fold, I see that a lot in other code bases.

`filter_map` is implemented like this:
```ocaml
let filter_map f =
  let rec aux accu = function
    | [] -> rev accu
    | x :: l ->
        match f x with
        | None -> aux accu l
        | Some v -> aux (v :: accu) l
  in
  aux []
```

Reply:
You're right!